File: phases/16-software-engineering-and-architecture/08-domain-driven-design-bounded-contexts-aggregates/code/main.ts
/**
 * Domain-Driven Design — Bounded Contexts, Aggregates
 * Phase 16 — Software Engineering & Architecture
 *
 * Order Management bounded context (TypeScript):
 *   - Value objects: Money, Address
 *   - Entities: OrderLine
 *   - Aggregate root: Order (enforces invariants)
 *   - Domain events: OrderPlaced, OrderCancelled
 *   - Repository: OrderRepository (collection-like interface)
 *   - Anti-Corruption Layer: OrderACL
 */

// ---------------------------------------------------------------------------
// Bounded Context Boundary: Order Management
// Everything below this line belongs to the Order Management context.
// Other contexts (Inventory, Billing) have their own models.
// ---------------------------------------------------------------------------

// --- Domain Events ---------------------------------------------------------

interface DomainEvent {
  readonly occurredAt: Date;
  readonly eventType: string;
}

interface OrderPlacedEvent extends DomainEvent {
  readonly eventType: "OrderPlaced";
  readonly orderId: string;
  readonly customerId: string;
  readonly totalAmount: number;
  readonly totalCurrency: Currency;
}

interface OrderCancelledEvent extends DomainEvent {
  readonly eventType: "OrderCancelled";
  readonly orderId: string;
  readonly reason: string;
}

type OrderEvent = OrderPlacedEvent | OrderCancelledEvent;

// --- Value Objects ---------------------------------------------------------

type Currency = "USD" | "EUR" | "GBP";

class Money {
  private constructor(
    public readonly amount: number,
    public readonly currency: Currency,
  ) {
    if (amount < 0) {
      throw new Error(`Money amount cannot be negative: ${amount}`);
    }
  }

  static of(amount: number, currency: Currency = "USD"): Money {
    return new Money(Math.round(amount * 100) / 100, currency);
  }

  add(other: Money): Money {
    if (this.currency !== other.currency) {
      throw new Error(`Cannot add ${this.currency} to ${other.currency}`);
    }
    return Money.of(this.amount + other.amount, this.currency);
  }

  multiply(factor: number): Money {
    return Money.of(
      Math.round(this.amount * factor * 100) / 100,
      this.currency,
    );
  }

  static zero(currency: Currency = "USD"): Money {
    return Money.of(0, currency);
  }

  equals(other: Money): boolean {
    return this.amount === other.amount && this.currency === other.currency;
  }

  toString(): string {
    return `${this.currency} ${this.amount.toFixed(2)}`;
  }
}

class Address {
  private constructor(
    public readonly street: string,
    public readonly city: string,
    public readonly state: string,
    public readonly postalCode: string,
    public readonly country: string,
  ) {
    if (!street.trim()) throw new Error("Street is required");
    if (!city.trim()) throw new Error("City is required");
    if (!country.trim()) throw new Error("Country is required");
  }

  static of(
    street: string,
    city: string,
    state: string,
    postalCode: string,
    country: string,
  ): Address {
    return new Address(street, city, state, postalCode, country);
  }

  formatted(): string {
    return `${this.street}, ${this.city}, ${this.state} ${this.postalCode}, ${this.country}`;
  }

  equals(other: Address): boolean {
    return (
      this.street === other.street &&
      this.city === other.city &&
      this.state === other.state &&
      this.postalCode === other.postalCode &&
      this.country === other.country
    );
  }
}

// --- Entities ---------------------------------------------------------------

class OrderLine {
  readonly lineId: string;
  readonly productId: string;
  readonly productName: string;
  quantity: number;
  readonly unitPrice: Money;

  constructor(
    lineId: string,
    productId: string,
    productName: string,
    quantity: number,
    unitPrice: Money,
  ) {
    if (quantity <= 0) {
      throw new Error(`Quantity must be positive, got ${quantity}`);
    }
    this.lineId = lineId;
    this.productId = productId;
    this.productName = productName;
    this.quantity = quantity;
    this.unitPrice = unitPrice;
  }

  lineTotal(): Money {
    return this.unitPrice.multiply(this.quantity);
  }

  changeQuantity(newQuantity: number): void {
    if (newQuantity <= 0) {
      throw new Error(`Quantity must be positive, got ${newQuantity}`);
    }
    this.quantity = newQuantity;
  }
}

// --- Aggregate Root ---------------------------------------------------------

type OrderStatus = "draft" | "placed" | "cancelled";

class Order {
  private readonly orderId: string;
  private readonly customerId: string;
  private shippingAddress: Address;
  private lines: OrderLine[];
  private status: OrderStatus;
  private events: OrderEvent[];

  constructor(orderId: string, customerId: string, shippingAddress: Address) {
    if (!customerId.trim()) {
      throw new Error("Customer ID is required");
    }
    this.orderId = orderId;
    this.customerId = customerId;
    this.shippingAddress = shippingAddress;
    this.lines = [];
    this.status = "draft";
    this.events = [];
  }

  get id(): string {
    return this.orderId;
  }

  get currentStatus(): OrderStatus {
    return this.status;
  }

  get allLines(): ReadonlyArray<OrderLine> {
    return this.lines;
  }

  get domainEvents(): ReadonlyArray<OrderEvent> {
    return [...this.events];
  }

  private ensureDraft(): void {
    if (this.status !== "draft") {
      throw new Error(
        `Cannot modify order in status '${this.status}'`,
      );
    }
  }

  private nextLineId(): string {
    return `LINE-${this.lines.length + 1}-${Date.now()}`;
  }

  addLine(
    productId: string,
    productName: string,
    quantity: number,
    unitPrice: Money,
  ): void {
    this.ensureDraft();

    // Invariant: no duplicate products
    const existing = this.lines.find((l) => l.productId === productId);
    if (existing) {
      throw new Error(
        `Product ${productId} already in order ${this.orderId}`,
      );
    }

    // Invariant: quantity must be positive
    if (quantity <= 0) {
      throw new Error(`Quantity must be positive, got ${quantity}`);
    }

    const line = new OrderLine(
      this.nextLineId(),
      productId,
      productName,
      quantity,
      unitPrice,
    );
    this.lines.push(line);
  }

  removeLine(productId: string): void {
    this.ensureDraft();
    const idx = this.lines.findIndex((l) => l.productId === productId);
    if (idx === -1) {
      throw new Error(
        `Product ${productId} not found in order ${this.orderId}`,
      );
    }
    this.lines.splice(idx, 1);
  }

  changeLineQuantity(productId: string, newQuantity: number): void {
    this.ensureDraft();
    const line = this.lines.find((l) => l.productId === productId);
    if (!line) {
      throw new Error(
        `Product ${productId} not found in order ${this.orderId}`,
      );
    }
    line.changeQuantity(newQuantity);
  }

  total(): Money {
    if (this.lines.length === 0) {
      return Money.zero();
    }
    return this.lines
      .map((l) => l.lineTotal())
      .reduce((acc, cur) => acc.add(cur));
  }

  place(): void {
    if (this.status !== "draft") {
      throw new Error(
        `Can only place a draft order, current status: ${this.status}`,
      );
    }
    if (this.lines.length === 0) {
      throw new Error("Cannot place an empty order");
    }

    this.status = "placed";
    const total = this.total();
    this.events.push({
      eventType: "OrderPlaced",
      occurredAt: new Date(),
      orderId: this.orderId,
      customerId: this.customerId,
      totalAmount: total.amount,
      totalCurrency: total.currency,
    });
  }

  cancel(reason: string): void {
    if (this.status !== "draft") {
      throw new Error(
        `Can only cancel a draft order, current status: ${this.status}`,
      );
    }
    this.status = "cancelled";
    this.events.push({
      eventType: "OrderCancelled",
      occurredAt: new Date(),
      orderId: this.orderId,
      reason,
    });
  }

  toString(): string {
    return `Order(id=${this.orderId}, customer=${this.customerId}, lines=${this.lines.length}, status=${this.status})`;
  }
}

// --- Repository -------------------------------------------------------------

class OrderRepository {
  private store: Map<string, Order> = new Map();

  add(order: Order): void {
    if (this.store.has(order.id)) {
      throw new Error(`Order ${order.id} already exists`);
    }
    this.store.set(order.id, order);
  }

  getById(orderId: string): Order | undefined {
    return this.store.get(orderId);
  }

  remove(orderId: string): void {
    if (!this.store.has(orderId)) {
      throw new Error(`Order ${orderId} not found`);
    }
    this.store.delete(orderId);
  }

  nextIdentity(): string {
    return crypto.randomUUID();
  }
}

// --- Anti-Corruption Layer --------------------------------------------------
// Translates data from an external (upstream) context into our model.

interface LegacyOrderDTO {
  rawOrderNum: string;
  rawCustomerRef: string;
  rawLines: Array<{
    sku: string;
    name?: string;
    qty: string;
    price: string;
  }>;
}

class OrderACL {
  static translate(dto: LegacyOrderDTO): Order {
    const order = new Order(
      `ORD-${dto.rawOrderNum}`,
      dto.rawCustomerRef,
      Address.of("Unknown", "Unknown", "N/A", "00000", "US"),
    );

    for (const rawLine of dto.rawLines) {
      order.addLine(
        rawLine.sku,
        rawLine.name ?? "Unknown Product",
        parseInt(rawLine.qty, 10),
        Money.of(parseFloat(rawLine.price)),
      );
    }

    return order;
  }
}

// --- Demo -------------------------------------------------------------------

function main(): void {
  const repo = new OrderRepository();

  // Create an order
  const orderId = repo.nextIdentity();
  const order = new Order(
    orderId,
    "CUST-42",
    Address.of("123 Elm St", "Springfield", "IL", "62704", "US"),
  );

  // Add lines through the aggregate root
  order.addLine("PROD-A", "Widget", 3, Money.of(9.99));
  order.addLine("PROD-B", "Gadget", 1, Money.of(49.99));

  console.log(`Order total: ${order.total().toString()}`);

  // Invariant: duplicate product rejected
  try {
    order.addLine("PROD-A", "Widget", 1, Money.of(9.99));
  } catch (e) {
    console.log(`Invariant enforced: ${(e as Error).message}`);
  }

  // Place the order — transitions state and publishes event
  order.place();
  console.log(`Order status: ${order.currentStatus}`);
  console.log(
    `Events published: ${JSON.stringify(order.domainEvents, null, 2)}`,
  );

  // Invariant: cannot modify a placed order
  try {
    order.removeLine("PROD-B");
  } catch (e) {
    console.log(`Invariant enforced: ${(e as Error).message}`);
  }

  // Persist via repository
  repo.add(order);

  // Anti-Corruption Layer: translate legacy data
  const legacyDTO: LegacyOrderDTO = {
    rawOrderNum: "LEG-007",
    rawCustomerRef: "EXT-CUST-99",
    rawLines: [
      { sku: "PROD-C", name: "Doohickey", qty: "2", price: "14.50" },
    ],
  };
  const translatedOrder = OrderACL.translate(legacyDTO);
  console.log(`ACL-translated order: ${translatedOrder.toString()}`);

  // Invariant: cannot cancel a placed order
  try {
    order.cancel("changed mind");
  } catch (e) {
    console.log(`Invariant enforced: ${(e as Error).message}`);
  }

  console.log(`Total lines in order: ${order.allLines.length}`);
}

main();

File: phases/16-software-engineering-and-architecture/08-domain-driven-design-bounded-contexts-aggregates/code/main.py
"""
Domain-Driven Design — Bounded Contexts, Aggregates
Phase 16 — Software Engineering & Architecture

Order Management bounded context:
  - Value objects: Money, Address
  - Entities: OrderLine
  - Aggregate root: Order (enforces invariants)
  - Domain events: OrderPlaced, OrderCancelled
  - Repository: OrderRepository (collection-like interface)
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from decimal import Decimal
from typing import List, Optional


# ---------------------------------------------------------------------------
# Bounded Context Boundary: Order Management
# Everything below this line belongs to the Order Management context.
# Other contexts (Inventory, Billing) have their own models.
# ---------------------------------------------------------------------------


# --- Domain Events ---------------------------------------------------------

@dataclass(frozen=True)
class DomainEvent:
    occurred_at: datetime


@dataclass(frozen=True)
class OrderPlaced(DomainEvent):
    order_id: str
    customer_id: str
    total_amount: Decimal
    total_currency: str


@dataclass(frozen=True)
class OrderCancelled(DomainEvent):
    order_id: str
    reason: str


# --- Value Objects ---------------------------------------------------------

@dataclass(frozen=True)
class Money:
    """
    A value object representing a monetary amount.
    Immutable — all operations return a new instance.
    Compared by value, not identity.
    """
    amount: Decimal
    currency: str

    def __post_init__(self):
        if self.currency not in ("USD", "EUR", "GBP"):
            raise ValueError(f"Unsupported currency: {self.currency}")
        if self.amount < 0:
            raise ValueError(f"Money amount cannot be negative: {self.amount}")

    def add(self, other: Money) -> Money:
        if self.currency != other.currency:
            raise ValueError(
                f"Cannot add {self.currency} to {other.currency}"
            )
        return Money(amount=self.amount + other.amount, currency=self.currency)

    def multiply(self, factor: Decimal) -> Money:
        return Money(amount=self.amount * factor, currency=self.currency)

    def zero(self) -> Money:
        return Money(amount=Decimal("0"), currency=self.currency)

    @classmethod
    def from_float(cls, amount: float, currency: str = "USD") -> Money:
        return cls(amount=Decimal(str(round(amount, 2))), currency=currency)


@dataclass(frozen=True)
class Address:
    """
    A value object representing a shipping address.
    Immutable — if the address changes, create a new instance.
    Compared by value, not identity.
    """
    street: str
    city: str
    state: str
    postal_code: str
    country: str

    def __post_init__(self):
        if not self.street.strip():
            raise ValueError("Street is required")
        if not self.city.strip():
            raise ValueError("City is required")
        if not self.country.strip():
            raise ValueError("Country is required")

    def formatted(self) -> str:
        return f"{self.street}, {self.city}, {self.state} {self.postal_code}, {self.country}"


# --- Entities ---------------------------------------------------------------

@dataclass
class OrderLine:
    """
    An entity within the Order aggregate.
    Has identity (line_id) that persists across state changes.
    Never referenced directly from outside the aggregate —
    all access goes through the Order aggregate root.
    """
    line_id: str
    product_id: str
    product_name: str
    quantity: int
    unit_price: Money

    def __post_init__(self):
        if self.quantity <= 0:
            raise ValueError(f"Quantity must be positive, got {self.quantity}")

    def line_total(self) -> Money:
        return self.unit_price.multiply(Decimal(self.quantity))

    def change_quantity(self, new_quantity: int) -> None:
        if new_quantity <= 0:
            raise ValueError(f"Quantity must be positive, got {new_quantity}")
        self.quantity = new_quantity


# --- Aggregate Root ---------------------------------------------------------

class Order:
    """
    Aggregate root for the Order Management bounded context.

    Invariants enforced:
      1. An order cannot contain duplicate lines for the same product.
      2. Quantities must be positive.
      3. Unit prices must be non-negative.
      4. A placed order cannot have lines added or removed.
      5. A cancelled order cannot be modified.

    All mutations go through the aggregate root.
    External references use order_id, not Order objects directly.
    """

    STATUS_DRAFT = "draft"
    STATUS_PLACED = "placed"
    STATUS_CANCELLED = "cancelled"

    def __init__(
        self,
        order_id: str,
        customer_id: str,
        shipping_address: Address,
    ):
        if not customer_id.strip():
            raise ValueError("Customer ID is required")
        self._order_id = order_id
        self._customer_id = customer_id
        self._shipping_address = shipping_address
        self._lines: List[OrderLine] = []
        self._status = self.STATUS_DRAFT
        self._events: List[DomainEvent] = []

    @property
    def order_id(self) -> str:
        return self._order_id

    @property
    def customer_id(self) -> str:
        return self._customer_id

    @property
    def shipping_address(self) -> Address:
        return self._shipping_address

    @property
    def status(self) -> str:
        return self._status

    @property
    def events(self) -> List[DomainEvent]:
        return list(self._events)

    def _ensure_draft(self) -> None:
        if self._status != self.STATUS_DRAFT:
            raise RuntimeError(
                f"Cannot modify order in status '{self._status}'"
            )

    def add_line(
        self,
        product_id: str,
        product_name: str,
        quantity: int,
        unit_price: Money,
    ) -> None:
        self._ensure_draft()

        # Invariant: no duplicate products
        if any(line.product_id == product_id for line in self._lines):
            raise ValueError(
                f"Product {product_id} already in order {self._order_id}"
            )

        # Invariant: quantity must be positive
        if quantity <= 0:
            raise ValueError(f"Quantity must be positive, got {quantity}")

        line = OrderLine(
            line_id=str(uuid.uuid4()),
            product_id=product_id,
            product_name=product_name,
            quantity=quantity,
            unit_price=unit_price,
        )
        self._lines.append(line)

    def remove_line(self, product_id: str) -> None:
        self._ensure_draft()
        if not any(line.product_id == product_id for line in self._lines):
            raise ValueError(
                f"Product {product_id} not found in order {self._order_id}"
            )
        self._lines = [l for l in self._lines if l.product_id != product_id]

    def change_line_quantity(self, product_id: str, new_quantity: int) -> None:
        self._ensure_draft()
        for line in self._lines:
            if line.product_id == product_id:
                line.change_quantity(new_quantity)
                return
        raise ValueError(
            f"Product {product_id} not found in order {self._order_id}"
        )

    def total(self) -> Money:
        if not self._lines:
            return Money(amount=Decimal("0"), currency="USD")
        result = self._lines[0].line_total()
        for line in self._lines[1:]:
            result = result.add(line.line_total())
        return result

    def place(self) -> None:
        if self._status != self.STATUS_DRAFT:
            raise RuntimeError(
                f"Can only place a draft order, current status: {self._status}"
            )
        if not self._lines:
            raise ValueError("Cannot place an empty order")
        self._status = self.STATUS_PLACED
        total = self.total()
        self._events.append(
            OrderPlaced(
                occurred_at=datetime.now(timezone.utc),
                order_id=self._order_id,
                customer_id=self._customer_id,
                total_amount=total.amount,
                total_currency=total.currency,
            )
        )

    def cancel(self, reason: str) -> None:
        if self._status != self.STATUS_DRAFT:
            raise RuntimeError(
                f"Can only cancel a draft order, current status: {self._status}"
            )
        self._status = self.STATUS_CANCELLED
        self._events.append(
            OrderCancelled(
                occurred_at=datetime.now(timezone.utc),
                order_id=self._order_id,
                reason=reason,
            )
        )

    def __repr__(self) -> str:
        return (
            f"Order(id={self._order_id}, customer={self._customer_id}, "
            f"lines={len(self._lines)}, status={self._status})"
        )


# --- Repository -------------------------------------------------------------

class OrderRepository:
    """
    Provides a collection-like interface for Order aggregates.
    Works with aggregate roots only — no OrderLineRepository.
    """

    def __init__(self) -> None:
        self._store: dict[str, Order] = {}

    def add(self, order: Order) -> None:
        if order.order_id in self._store:
            raise ValueError(
                f"Order {order.order_id} already exists"
            )
        self._store[order.order_id] = order

    def get_by_id(self, order_id: str) -> Optional[Order]:
        return self._store.get(order_id)

    def remove(self, order_id: str) -> None:
        if order_id not in self._store:
            raise ValueError(f"Order {order_id} not found")
        del self._store[order_id]

    def next_identity(self) -> str:
        return str(uuid.uuid4())


# --- Anti-Corruption Layer --------------------------------------------------
# Translates data from an external (upstream) context into our model.

class LegacyOrderDTO:
    """Represents data coming from an upstream legacy system."""
    raw_order_num: str
    raw_customer_ref: str
    raw_lines: List[dict]

    def __init__(
        self,
        raw_order_num: str,
        raw_customer_ref: str,
        raw_lines: List[dict],
    ):
        self.raw_order_num = raw_order_num
        self.raw_customer_ref = raw_customer_ref
        self.raw_lines = raw_lines


class OrderACL:
    """
    Anti-Corruption Layer: translates legacy data into the
    Order Management bounded context's model, protecting our
    invariants from upstream schema changes.
    """

    @staticmethod
    def translate(dto: LegacyOrderDTO) -> Order:
        order = Order(
            order_id=f"ORD-{dto.raw_order_num}",
            customer_id=dto.raw_customer_ref,
            shipping_address=Address(
                street="Unknown",
                city="Unknown",
                state="N/A",
                postal_code="00000",
                country="US",
            ),
        )
        for raw_line in dto.raw_lines:
            order.add_line(
                product_id=raw_line["sku"],
                product_name=raw_line.get("name", "Unknown Product"),
                quantity=int(raw_line["qty"]),
                unit_price=Money.from_float(float(raw_line["price"])),
            )
        return order


# --- Demo -------------------------------------------------------------------

def main() -> None:
    repo = OrderRepository()

    # Create an order
    order_id = repo.next_identity()
    order = Order(
        order_id=order_id,
        customer_id="CUST-42",
        shipping_address=Address(
            street="123 Elm St",
            city="Springfield",
            state="IL",
            postal_code="62704",
            country="US",
        ),
    )

    # Add lines through the aggregate root
    order.add_line(
        product_id="PROD-A",
        product_name="Widget",
        quantity=3,
        unit_price=Money(Decimal("9.99"), "USD"),
    )
    order.add_line(
        product_id="PROD-B",
        product_name="Gadget",
        quantity=1,
        unit_price=Money(Decimal("49.99"), "USD"),
    )

    print(f"Order total: {order.total()}")

    # Invariant: duplicate product rejected
    try:
        order.add_line(
            product_id="PROD-A",
            product_name="Widget",
            quantity=1,
            unit_price=Money(Decimal("9.99"), "USD"),
        )
    except ValueError as e:
        print(f"Invariant enforced: {e}")

    # Place the order — transitions state and publishes event
    order.place()
    print(f"Order status: {order.status}")
    print(f"Events published: {order.events}")

    # Invariant: cannot modify a placed order
    try:
        order.remove_line("PROD-B")
    except RuntimeError as e:
        print(f"Invariant enforced: {e}")

    # Persist via repository
    repo.add(order)

    # Anti-Corruption Layer: translate legacy data
    legacy_dto = LegacyOrderDTO(
        raw_order_num="LEG-007",
        raw_customer_ref="EXT-CUST-99",
        raw_lines=[
            {"sku": "PROD-C", "name": "Doohickey", "qty": "2", "price": "14.50"},
        ],
    )
    translated_order = OrderACL.translate(legacy_dto)
    print(f"ACL-translated order: {translated_order}")

    # Invariant: cannot cancel a placed order
    try:
        order.cancel("changed mind")
    except RuntimeError as e:
        print(f"Invariant enforced: {e}")

    print(f"Total lines in order: {len(order._lines)}")


if __name__ == "__main__":
    main()